### PR TITLE
Caff node force inclusion checker

### PR DESCRIPTION
**Pull request by ImJeremyHe**
### This PR:
- Adds a force inclusion checker in caff node following [this design doc](https://github.com/EspressoSystems/book/pull/52). The only change is getting the delayed count using the existing function.
- Adds a test

### Key places to review:
- caff_node_test.go

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1209845224208640
  - https://app.asana.com/0/0/1209837645605781

### CODE REVIEW

File: arbnode/espresso_force_inclusion_checker.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	if err != nil {
+		return 0
+	}
+	currentParentChainBlock, err := f.l1Reader.Client().BlockByNumber(ctx, nil)
+	if err != nil {
+		return 0
+	}
+
+	lastBadBlockNumber := arbmath.SaturatingUSub(f.config.BlockThresholdTolerance+currentParentChainBlock.NumberU64(), arbmath.BigToUintSaturating(maxTimeVariationDelayBlocks))
+	lastBadBlockTime := arbmath.SaturatingUSub(f.config.SecondThresholdTolerance+currentParentChainBlock.Time(), arbmath.BigToUintSaturating(maxTimeVariationDelaySeconds))
```
Comment (zacshowa) [inline arbnode/espresso_force_inclusion_checker.go]: Are we sure that these don't over/underflow? I see we are combining raw values from the config while doing some saturating subs. 
Just trying to make sure we cover our edge cases.

**Review by zacshowa — COMMENTED** [review]
This looks like a great start to this PR to me. I think it would be nice to have a full integration test that sees the caff node panic when there is a risk of force inclusion. However, I'm pretty confident in the code here after reading it/ seeing the unit test.

**Review by Sneh1999 — COMMENTED** [review]
@ImJeremyHe  can you please split the statechecker and forceInclusionChecker into two separate PR's? Because they are two different asana tickets. Also its hard to review both in one PR

File: arbnode/espresso_watchdog.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+	PollingInterval: time.Second * 100,
+}
+
+func EspressoStateCheckerConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Bool(prefix+".enable", DefaultStateCheckerConfig.Enable, "enable state checker")
+	f.Duration(prefix+".polling-interval", DefaultStateCheckerConfig.PollingInterval, "time after a success")
+	f.String(prefix+".trusted-node-url", DefaultStateCheckerConfig.TrustedNodeUrl, "http endpoint of the trusted node")
+}
+
+type StateChecker struct {
```
Comment (Sneh1999) [inline arbnode/espresso_watchdog.go]: we should put this in a separate file

File: arbnode/espresso_watchdog.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+func (f *ForceInclusionChecker) findFirstParentChainBlockBelow(ctx context.Context, lastBadBlockNumber uint64, lastBadBlockTime uint64) uint64 {
+	client := f.l1Reader.Client()
+	blockNumber := lastBadBlockNumber
+
+	for blockNumber > 0 {
+		block, err := client.BlockByNumber(ctx, arbmath.UintToBig(blockNumber))
+		if err != nil {
+			return 0
+		}
+		if block.NumberU64() <= lastBadBlockNumber || block.Time() <= lastBadBlockTime {
```
Comment (Sneh1999) [inline arbnode/espresso_watchdog.go]: should this not be `block.NumberU64() <= lastBadBlockNumber && block.Time() <= lastBadBlockTime`?

File: arbnode/espresso_watchdog.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+	maxTimeVariationDelayBlocks, _, maxTimeVariationDelaySeconds, _, err := f.seqInbox.MaxTimeVariation(ctx)
+	if err != nil {
+		return 0
+	}
+	currentParentChainBlock, err := f.l1Reader.Client().BlockByNumber(ctx, nil)
+	if err != nil {
+		return 0
+	}
+
+	lastBadBlockNumber := arbmath.SaturatingUSub(f.config.BlockThresholdTolerance+currentParentChainBlock.NumberU64(), arbmath.BigToUintSaturating(maxTimeVariationDelayBlocks))
```
Comment (Sneh1999) [inline arbnode/espresso_watchdog.go]: The code should work for orbit chains, 
if the l1Reader is arbitrum we need to still use ethereum blocks because `maxTimeVariationDelayBlocks` is determined by ethereum blocks not arbitrum blocks. Check this function:
https://github.com/OffchainLabs/arbitrum-sdk/blob/792a7ee3ccf09842653bc49b771671706894cbb4/src/lib/inbox/inbox.ts#L87C17-L87C36

**Review by Sneh1999 — COMMENTED** [review]
The way this code is written wont work because you are comparing the totalDelayedMessages at the current arbitrum block number with the `delayedMessageCount` lastBadBlockNumber arbitrum. 

But maxTimeVariationDelayBlocks and timestamp is wrt to ethereum not arbitrum, so we will have to fix that for orbit chains

File: arbnode/espresso_watchdog.go (hunk 3 above)
Comment (ImJeremyHe) [reply arbnode/espresso_watchdog.go]: Sorry for the late reply.

I think it should be `or` because that's[ what batcher does](https://github.com/EspressoSystems/nitro-espresso-integration/blob/161b3bbe3bf083f4ea4175c71d5ead858f4876bd/arbnode/batch_poster.go#L1639)

File: arbnode/espresso_force_inclusion_checker.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+	seqInbox SeqInboxInterface,
+	config ForceInclusionCheckerConfig,
+	l1Reader *headerreader.HeaderReader,
+	delayedMessageFetcher *DelayedMessageFetcher,
+	fatalErrChan chan error,
+) *ForceInclusionChecker {
+	return &ForceInclusionChecker{
+		seqInbox:              seqInbox,
+		config:                config,
+		l1Reader:              l1Reader,
```
Comment (Sneh1999) [inline arbnode/espresso_force_inclusion_checker.go:90]: can l1Reader be nil here?

**Review by Sneh1999 — COMMENTED** [review]
@ImJeremyHe I think you still didnt covert this case:
https://github.com/OffchainLabs/arbitrum-sdk/blob/792a7ee3ccf09842653bc49b771671706894cbb4/src/lib/inbox/inbox.ts#L104-L113

File: arbnode/espresso_watchdog.go (hunk 3 above)
Comment (Sneh1999) [reply arbnode/espresso_watchdog.go]: @ImJeremyHe  I was reading arbitrum's code and it looks like they are doing something similar to && here
https://github.com/OffchainLabs/arbitrum-sdk/blob/792a7ee3ccf09842653bc49b771671706894cbb4/src/lib/inbox/inbox.ts#L123-L125

Let me know if that is not true!

File: arbnode/espresso_force_inclusion_checker.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	if f.l1Reader.IsParentChainArbitrum() {
+		n, err := node_interfacegen.NewNodeInterface(types.NodeInterfaceAddress, f.l1Reader.Client())
+		if err != nil {
+			return 0, err
+		}
+		rng, err := n.L2BlockRangeForL1(&bind.CallOpts{Context: ctx}, lastBadBlockNumber)
+		if err != nil {
+			return 0, err
+		}
+		lastBadBlockNumber = rng.LastBlock
```
Comment (Sneh1999) [inline arbnode/espresso_force_inclusion_checker.go]: why are we only updating the block number and not the block timestamp as well?

Comment (ImJeremyHe) [reply arbnode/espresso_force_inclusion_checker.go]: Block number needs to be updated because of different speed of block creation in different chains. I think an accurate way is like this:
- Get current L2 block
- Find the corresponding eth timestamp/block number
- Calculate the dangerous eth timestamp/block number
- Get the corresponding L2 timestamp/block number
- Find the L2 block

For timestamp, it won't be huge difference with:
- Get current L2 block
- Calculate the dangerous timestamp
- Find the L2 block

Arbitrum sdk does similar things:
https://github.com/OffchainLabs/arbitrum-sdk/blob/792a7ee3ccf09842653bc49b771671706894cbb4/src/lib/inbox/inbox.ts#L233

File: arbnode/espresso_force_inclusion_checker.go (hunk 5 above)
Comment (ImJeremyHe) [reply arbnode/espresso_force_inclusion_checker.go:90]: No. Caff node requires l1 reader to get delayed messages. Force inclusion checker uses the same l1 reader

**Review by Sneh1999 — COMMENTED** [review]
@ImJeremyHe  for some reason all tests are failing 🤔

File: arbnode/espresso_force_inclusion_checker.go (hunk 6 above)
Comment (Sneh1999) [reply arbnode/espresso_force_inclusion_checker.go]: okay makes sense!

File: arbnode/espresso_utils.go [hunk 7]
```diff
@@ -15,6 +16,38 @@ import (
...
+
+// Looks for the first block number that is equal to or greater than the target
+func binarySearchForBlockNumber(
+	ctx context.Context,
+	start, end uint64,
+	f func(context.Context, uint64) (int, error),
+) (uint64, error) {
+	for start < end {
+		mid := (start + end) / 2
+		fmt.Println("mid", mid)
```
Comment (Sneh1999) [inline arbnode/espresso_utils.go]: nit: we should remove this

File: arbnode/espresso_utils.go [hunk 8]
```diff
@@ -15,6 +16,38 @@ import (
...
+			return 0, err
+		}
+		if result == binarySearch_GreaterThanTarget {
+			end = mid
+		} else if result == binarySearch_LessThanTarget {
+			start = mid + 1
+		} else {
+			// We are looking for the first block number.
+			// So the loop should continue until start == end
+			end = mid
```
Comment (Sneh1999) [inline arbnode/espresso_utils.go:44]: I didnt understand what this means? why are we setting end = mid here?

File: arbnode/espresso_force_inclusion_checker.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+	if f.l1Reader.IsParentChainArbitrum() {
+		n, err := node_interfacegen.NewNodeInterface(types.NodeInterfaceAddress, f.l1Reader.Client())
+		if err != nil {
+			return 0, err
+		}
+		rng, err := n.L2BlockRangeForL1(&bind.CallOpts{Context: ctx}, lastBadBlockNumber)
+		if err == nil {
+			lastBadBlockNumber = rng.LastBlock
+		} else {
+			genesis, err := n.NitroGenesisBlock(&bind.CallOpts{Context: ctx})
```
Comment (Sneh1999) [inline arbnode/espresso_force_inclusion_checker.go:186]: can you add a comment why we are doing this?

File: arbnode/espresso_utils.go (hunk 8 above)
Comment (ImJeremyHe) [reply arbnode/espresso_utils.go:44]: Let's say we want the first L2 block number for the L1 number 70.

Generally, there will be a range of L2 blocks. Let's say 100-120

When we check the `mid` which is 110, we should not stop because we are not sure if it is the first.

So we should continue the proccess until `start==end`

**Review by Sneh1999 — APPROVED** [review]
LGTM!
